web/src/scripts/components/publishing/PublishingMetadata.jsx

```diff
@@ -0,0 +1,390 @@
+      case PrimitiveTypes.RAW: return ''
+    }
+  }
+
+  return null
+}
+
+export default class extends Component {
+  constructor(props) {
+    super()
+
+    const {component} = props
+
+    this.state = {
+      component: _.cloneDeep(component),
+    }
+  }
+
+  renderCode(component) {
+    const {tagName, props, dependencies, imports} = component
+
+    const chunks = [
+      _.map(imports, (value, key) => {
+        if (_.isString(value)) {
+          return `import ${value} from '${key}'`
+        } else if (_.isArray(value)) {
+          return `import { ${value.join(', ')} } from '${key}'`
+        }
+      }).join('\n'),
+      createJSX({tagName, props}),
+    ]
+
+    return (
+      <pre style={styles.code} key={'code'}>{chunks.join('\n\n')}</pre>
+    )
+  }
+
+  renderMap(map, mapName) {
+    const {component} = this.state
+    const {width} = this.props
+
+    return _.map(map, (value, key) => {
+      let inputElement
+
+      if (_.isString(value)) {
+        inputElement = (
+          <StringInput
+            value={value}
+            onChange={(newValue) => this.updateComponent(c => c[mapName][key] = newValue)}
+          />
+        )
+      } else if (_.isArray(value)) {
+        inputElement = (
+          <StringInput
+            value={value.join(',')}
+            onChange={(newValue) => {
+              this.updateComponent(c => {
+                try {
+                  const parts = newValue.split(',')
+                  c[mapName][key] = parts
+                } catch (e) {
+                  console.log('failed to change', value, '=>', newValue)
+                }
+              })
+            }}
+          />
+        )
+      }
+
+      return (
+        <InspectorField
+          key={key}
+          name={key}
+          inset={INSET_WIDTH}
+          width={width}
+          deletable={true}
+          inputElement={inputElement}
+          menuElement={(
+            <NameEditor
+              name={key}
+              withConfirmation={true}
+              onChange={(newKey) => {
+                this.updateComponent(c => {
+                  delete c[mapName][key]
+                  c[mapName][newKey] = value
+                })
+              }}
+            />
+          )}
+          onDelete={() => this.updateComponent(c => delete c[mapName][key])}
+        />
+      )
+    })
+  }
+
+  updateComponent(updater) {
+    const {onUpdateComponent} = this.props
+    const {component} = this.state
+    const clone = _.cloneDeep(component)
+    updater(clone)
+    this.setState({component: clone})
+    onUpdateComponent(clone)
+  }
+
+  render() {
+    const {width, onDeleteComponent} = this.props
+    const {component} = this.state
+    const {props: componentProps = [], dependencies, imports} = component
+
+    const rowDimensions = {
+      labelWidth: width - INPUT_WIDTH - 10,
+      inset: INSET_WIDTH,
+      inputWidth: INPUT_WIDTH,
+    }
+
+    return (
```

Comment:
Will refactor after I get designs from Roy.

I can redact the component from the PR if you like. You just won't be able to test publishing. But it doesn't work that well yet anyway, so maybe that's the right move.